matrix/matrix.py

```diff
@@ -0,0 +1,198 @@
+import operator
+from itertools import repeat
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type. This
+    implementation only supports even-side matrices (n * n).
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+                         the matrix is represented by a tuple.
+    """
+    __slots__ = '_matrix'
+
+    def __init__(self, *lines):
+        if not self._is_valid_data(*lines):
+            raise ValueError('Data is not valid')
+
+        super(Matrix, self).__setattr__('_matrix', tuple(lines))
+
+    def __setattr__(self, key, value):
```

Comment:
This defense is not needed
If someone wants to change the value he can, for example, override this function

You just need to make it not too easy for him, e.g. using tuples instead of lists